cumulus/client/consensus/aura/src/collators/slot_based/slot_timer.rs

```diff
@@ -0,0 +1,266 @@
+	let mut block_production_interval = min(para_slot_duration.as_duration(), relay_slot_duration);
+
+	if assigned_core_num > para_slots_per_relay_block &&
+		para_slot_duration.as_duration() >= relay_slot_duration
+	{
+		block_production_interval =
+			max(relay_slot_duration / assigned_core_num, BLOCK_PRODUCTION_MINIMUM_INTERVAL_MS);
+		tracing::debug!(
+			target: LOG_TARGET,
+			?block_production_interval,
+			"Expected to produce for {assigned_core_num} cores but only have {para_slots_per_relay_block} slots. Attempting to produce multiple blocks per slot."
+		);
+	}
+
+	let (duration, timestamp) = time_until_next_attempt(time_now, block_production_interval, time_offset);
+	let aura_slot = Slot::from_timestamp(timestamp, para_slot_duration);
+	(duration, timestamp, aura_slot)
+}
+
+/// Returns current duration since Unix epoch.
+fn duration_now() -> Duration {
+	use std::time::SystemTime;
+	let now = SystemTime::now();
+	now.duration_since(SystemTime::UNIX_EPOCH).unwrap_or_else(|e| {
+		panic!("Current time {:?} is before Unix epoch. Something is wrong: {:?}", now, e)
+	})
+}
+
+/// Returns the duration until the next block production should be attempted.
+/// Returns:
+/// - Duration: The duration until the next attempt.
+/// - Timestamp: The time at which the attempt will take place.
+fn time_until_next_attempt(
+	now: Duration,
+	block_production_interval: Duration,
+	offset: Duration,
+) -> (Duration, Timestamp) {
+	let now = now.as_millis().saturating_sub(offset.as_millis());
+
+	let next_slot_time = ((now + block_production_interval.as_millis()) /
+		block_production_interval.as_millis()) *
+		block_production_interval.as_millis();
+	let remaining_millis = next_slot_time - now;
+	(Duration::from_millis(remaining_millis as u64), Timestamp::from(next_slot_time as u64))
+}
+
+impl<Block, Client, P> SlotTimer<Block, Client, P>
+where
+	Block: BlockT,
+	Client: ProvideRuntimeApi<Block> + Send + Sync + 'static + UsageProvider<Block>,
+	Client::Api: AuraApi<Block, P::Public>,
+	P: Pair,
+	P::Public: AppPublic + Member + Codec,
+	P::Signature: TryFrom<Vec<u8>> + Member + Codec,
+{
+	/// Create a new slot timer.
+	pub fn new_with_offset(
+		client: Arc<Client>,
+		time_offset: Duration,
+		relay_slot_duration: Duration,
+	) -> Self {
+		Self {
+			client,
+			time_offset,
+			last_reported_core_num: None,
+			relay_slot_duration,
+			_marker: Default::default(),
+		}
+	}
+
+	/// Inform the slot timer about the last seen number of cores.
+	pub fn update_scheduling(&mut self, num_cores_next_block: u32) {
+		self.last_reported_core_num = Some(num_cores_next_block);
+	}
+
+	/// Returns a future that resolves when the next block production should be attempted.
+	pub async fn wait_until_next_slot(&self) -> Result<SlotInfo, ()> {
+		let Ok(slot_duration) = crate::slot_duration(&*self.client) else {
+			tracing::error!(target: LOG_TARGET, "Failed to fetch slot duration from runtime.");
+			return Err(())
```

Comment:
The function could just return an `Option`?

Reply:
It could, but IMO this is an error where we failed to fetch the slot duration.